In [1]:
#import plotly.graph_objects as go
from plotly import graph_objects as go
import plotly.express as px
import matplotlib.pyplot as plt
import numpy as np
from sklearn import datasets
import sympy as sym
from sympy import symbols
import pandas as pd

In [2]:
#%store -r Mmax
#%store -r Mmin 
#%store -r MEd_slu
#MEd_slu 

# Mmax = np.loadtxt('/home/matteo/ownCloud/UNITN/Tecnica-delle-Costruzioni/Esercitazione/TdC/export/var/Mmax.txt') #Linux
Mmax = np.loadtxt('C:\\Users\\anzo5\\ownCloud\\UNITN\\Tecnica-delle-Costruzioni\\Esercitazione\\TdC\\export\\var\\Mmax.txt') #Windows
# Mmin = np.loadtxt('/home/matteo/ownCloud/UNITN/Tecnica-delle-Costruzioni/Esercitazione/TdC/export/var/Mmin.txt') #Linux
Mmin = np.loadtxt('C:\\Users\\anzo5\\ownCloud\\UNITN\\Tecnica-delle-Costruzioni\\Esercitazione\\TdC\\export\\var\\Mmin.txt')#Windows
# MEd_slu = pd.read_csv('/home/matteo/ownCloud/UNITN/Tecnica-delle-Costruzioni/Esercitazione/TdC/export/var/MEd_slu.csv') #Linux
MEd_slu = pd.read_csv('C:\\Users\\anzo5\\ownCloud\\UNITN\\Tecnica-delle-Costruzioni\\Esercitazione\\TdC\\export\\var\\MEd_slu.csv')#Windows
# MEd_slu.to_latex('MEd_slu.tex')
MEd_slu

Section  MEd+ [kN m]  smax [m]  MEd- [kN m]  smin [m]
0       C1    95.189448      1.26          NaN       NaN
1       N2          NaN       NaN  -199.249367      3.00
2       C2   156.565614      5.32     0.000000       NaN
3       N3          NaN       NaN  -216.792778      7.50
4       C3   124.347537      9.55   -39.690612      9.47
5       N4          NaN       NaN  -210.348968     11.50
6       C4   151.829401     13.96    -6.190556     13.74
7       N5          NaN       NaN  -236.629193     16.50
8       C5   144.511198     19.56     0.000000       NaN
9       N6          NaN       NaN  -187.959174     22.65
10      C6    87.580801     25.00          NaN       NaN

In [3]:
smax = MEd_slu[['smax [m]']].to_numpy()
smin = MEd_slu[['smin [m]']].to_numpy()

smaxmin = np.sort((np.concatenate((smax, smin))).reshape(-1))
smaxmin = np.insert(smaxmin, -1, 26.65)

#ss is for plot
ss=smaxmin
for i in range (0, len(ss)-1):
    if ss[i+1] - ss[i] < 0.5:
        ss[i] = ss[i+1]
        
ss

array([ 1.26,  3.  ,  5.32,  7.5 ,  9.55,  9.55, 11.5 , 13.96, 13.96,
       16.5 , 19.56, 22.65, 25.  ,   nan,   nan,   nan,   nan,   nan,
         nan,   nan,   nan, 26.65,   nan])

In [5]:
s = np.arange(0,26.65, .001)

MR1t = 50* np.ones(len(s))

fig = go.Figure()

fig.add_trace(go.Scatter(
    x = s,
    y = Mmax,
    name='M+',
    line_color='#000000',
    showlegend=False,
    line = dict(width=1)
))

fig.add_trace(go.Scatter(
    x = s,
    y = Mmin,
    name='M-',
    line_color='#000000',
    fill='tonexty',
    showlegend=False,
    line = dict(width=1)
))

# fig.add_trace(go.Scatter(
#     x = s, 
#     y = MR1t,
#     name='2ϕ18',
#     showlegend=True,
#     line_color='#90ee90'
# ))

# fig.add_trace(go.Scatter(
#     x = s,
#     y = -MR1t,
#     name = '2ϕ18',
#     #sholegend = True,
#     line_color = '#90ee90'
# ))

fig.update_yaxes(autorange="reversed")
fig.update_traces(mode='lines')
fig.update_layout(
    title_text=('Inviluppo dei momenti flettenti<br>' + 'Combinazione SLU'),
    title_x=.5,
    autosize=True,
    xaxis = dict(
        tickmode='array',
        tickvals=ss,
        ticktext=ss,
        title=r'$s\,[m]$',
    ),
    yaxis = dict(
        tickmode = 'linear',
        tick0 = 0,
        dtick = 50,
        title=r'$M_{Ed}\,[kN\,m]$'
    ),
    xaxis_showgrid=True, 
    yaxis_showgrid=True,
)

fig.write_image("Esercitazione/TdC/LaTeX/img/bendingMomentEnvelope.eps")

fig.update_yaxes(automargin=True)
fig.show()

NoSuchProcess: psutil.NoSuchProcess no process found with pid 28084

In [ ]:
0.9/2 *460*(2.5-0)

In [6]:
np.array([s2[1260], s1[1260]])

NameError: name 's2' is not defined

In [ ]:
sc1 = np.array([s2[1260], s1[1260]])
Mc1_max = Mmax[1260]*np.ones(len(sc1))

sc2 = np.array([s2[5320], s1[5320]])
Mc2_max = Mmax[5320]*np.ones(len(sc2))


In [ ]:

fig = go.Figure()

fig.add_trace(go.Scatter(
    x = sc1,
    y = Mc1_max,
    line_color= 'red',
    fillcolor='red',
    mode = 'none',
    fill='tozeroy',
    showlegend=False,
    line = dict(width=1)
))

fig.add_trace(go.Scatter(
    x = sc2,
    y = Mc2_max,
    line_color= 'red',
    fillcolor='red',
    mode = 'none',
    fill='tozeroy',
    showlegend=False,
    line = dict(width=1)
))

fig.add_trace(go.Scatter(
    x = s,
    y = Mmax,
    name='M+',
    line_color='#000000',
    showlegend=False,
    line = dict(width=1)
))



fig.add_trace(go.Scatter(
    x = s2,
    y = Mmax,
    name='M+',
    line_color= 'red',
    fillcolor='red',
    mode = 'none',
    fill='tozeroy',
    showlegend=False,
    line = dict(width=1)
))



# fig.add_trace(go.Scatter(
#     x = s,
#     y = MR1t,
#     name='2ϕ18',
#     showlegend=True,
#     line_color='#90ee90'
# ))

# fig.add_trace(go.Scatter(
#     x = s,
#     y = -MR1t,
#     name = '2ϕ18',
#     #sholegend = True,
#     line_color = '#90ee90'
# ))

fig.update_yaxes(autorange="reversed")
fig.update_traces(mode='lines')
fig.update_layout(
    title_text=('Inviluppo dei momenti flettenti<br>' + 'Combinazione SLU'),
    title_x=.5,
    autosize=True,
    xaxis = dict(
        tickmode='array',
        tickvals=ss,
        ticktext=ss,
        title=r'$s\,[m]$',
    ),
    yaxis = dict(
        tickmode = 'linear',
        tick0 = 0,
        dtick = 50,
        title=r'$M_{Ed}\,[kN\,m]$'
    ),
    xaxis_showgrid=True, 
    yaxis_showgrid=True,
)

fig.update_yaxes(automargin=True)
fig.show()

In [ ]:
s1[-1]

In [ ]:
s1 = s+ 517.5/1000
smax1 = smax + 517.5/1000
for i in range (0, len(s)):
    if s1[i] > 26.65:
        s1[i] = 26.65

In [7]:
s2 = s- 517.5/1000
smax2 = smax - 517.5/1000
for i in range (0, len(s)):
    if s2[i] < 0:
        s2[i] = 0

---

## Progetto sezione N5

Si ipotizzi di essere sulla retta limite tra il campo 2 e il campo 3. In questo modo:
$$
\xi_{2/3} = \dfrac{\varepsilon_{cu}}{\varepsilon_{cu} + \varepsilon_{su}}
$$

Assumendo 
\begin{align*}
\varepsilon_{cu} &= 0.35\%\\
\varepsilon_{su} &= 1.0\%
\end{align*}
risulta
$$
\xi_{2/3} = 0.2593
$$

Essendo sulla retta limite tra i campi 2 e 3, i valori di $\psi$ e $\lambda$ sono costanti e valgono
\begin{align*}
\psi &= 0.80952\\
\lambda &= 0.416
\end{align*}

In [131]:
b, psi, xi, d, fcd, beta, As, fyd, d2, MEd, mylambda, x, A1s, Es, epsilon_su = symbols('b psi xi d fcd beta As fyd d2 MEd mylambda x A1s Es, epsilon_su')
type(d)

sympy.core.symbol.Symbol

In [132]:
MEd_N5 = 234.169 #kN m

In [133]:
eq = b*psi*xi*d**2 * fcd* (1-mylambda) + beta* (b*psi*xi*d*fcd / (1-beta)) * (d-d2) - MEd
type(eq)

sympy.core.add.Add

In [134]:
eq.subs({b:300, psi:0.806399941030868, xi:0.256147482947766, fcd:14.17, beta:0.3, mylambda:0.41481981971363, fyd:391.3, d2:40, MEd:234.169*10**6})

513.832172638218*d**2 + 376.317920005007*d*(d - 40) - 234169000.0

In [135]:
num_eq = eq.subs({b:300, psi:0.80952, xi:0.2593, fcd:14.17, beta:0.3, mylambda:0.416, fyd:391.3, d2:40, MEd:234.169*10**6})

In [136]:
d_sol=sym.solve(num_eq, d)
d_sol[1]

517.621512774738

In [137]:
sol_As = b*psi*xi*d*fcd / (fyd*(1-beta))
sol_As

b*d*fcd*psi*xi/(fyd*(1 - beta))

In [138]:
num_As = sol_As.subs({b:300, psi:0.80952, xi:0.2593, fcd:14.17, beta:0.3, fyd:391.3, d:d_sol[1]})
num_As

1686.26425634593

In [139]:
num_A1s = .3 * num_As
num_A1s

505.879276903778

Adottando
\begin{align}
d &= 460\,mm\\
\end{align}
si ottiene

In [140]:
d_sol1=460 #è verificata anche con d = 460mm

In [141]:
num1_As = sol_As.subs({b:300, psi:0.80952, xi:0.2593, fcd:14.17, beta:0.3, fyd:391.3, d:d_sol1})
num1_As

1498.54969079829

In [142]:
num1_A1s = .3 * num1_As
num1_A1s

449.564907239487

Si sceglie allora
\begin{align}
A_s &= 6\Phi\,18mm = 486\pi\,mm^2\\
A'_s &= 2\Phi\,18mm = 162\pi\,mm^2
\end{align}

In [143]:
As_sol1 = np.pi*6*18**2 /4 #mm**2
A1s_sol1 = np.pi*2*18**2 /4 #mm**2

## Verifica sezione N5
Se ipotizziamo di essere in campo 3, $\psi$ e $\lambda$ sono costanti. Il copriferro superiore al limite elastico tra i campi 2 e 3 vale
$$
d''_{\varepsilon'_{2/3}} = \dfrac{\varepsilon_{cu} - \varepsilon'_{se}}{\varepsilon_{cu}}\,x_{2/3} = \dfrac{0.35-0.1863}{0.35}\%\,\cdot0.2593 \cdot 560\,mm = 55.788\,mm 
$$

Poiché $d'' = 40\,mm < d''_{\varepsilon'_{2/3}} = 67.916\,mm$ le armature superiori snervano in campo 2, sottocampo 2b).

Allora
$$
b\,\psi\,x\,fcd + A'_s\,f'_{yd} - A_s\,f_{yd} = 0
$$

da cui si ricava $x$

In [144]:
eq2 = b*psi*x*fcd+A1s*fyd-As*fyd
sol_eq2 = eq2.subs({b:300, psi:0.80952, fcd:14.17, A1s:A1s_sol1, As:As_sol1, fyd:391.3})
x2 = sym.solve(sol_eq2, x)
sol_eq2, x2
x2 = x2[0]
xi2 = x2/d_sol1
print(xi2,'< 0.2593')
xi2 < 0.2593

0.251610181635942 < 0.2593


True

L'ipotesi di campo 3 è **errata**!

Si ipotizza di essere in campo 2. Allora $\{\psi, \lambda\} \neq cost$.

**HP: sottocampo 2b ($\sigma'_s = f_{yd}$)**

Se $\varepsilon_c \in [\varepsilon_{c2}, \varepsilon_{cu}]$ allora
$$
\psi = 1 - \dfrac{1}{3}\,\dfrac{\varepsilon_{c2}}{\varepsilon_{c}} = 1 - \dfrac{1}{3}\,\dfrac{\varepsilon_{c2}}{\varepsilon_{su}}\,\dfrac{1-\xi}{\xi} = 1 - \dfrac{1}{3}\,\dfrac{2}{10}\cdot\dfrac{1-\xi}{\xi} = \frac{16 \xi - 1}{15 \xi}
$$

In [145]:
sym.simplify(1-(1/3) * (2/10) * (1-xi)/xi, rational=True)

(16*xi - 1)/(15*xi)

Essendo per ipotesi in campo 2b, le equazioni di equilibrio risultano
\begin{align}
b\,\psi\,\dfrac{x}{d}\,d\,f_{cd} + A'_s\,f_{yd} - A_s\,f_{yd} &= 0
\end{align}
cioè
$$
b\,\psi\,\xi\,d\,f_{cd} = f_{yd}\,(A_s - A'_s)
$$
ovvero
$$ 
\psi\,\xi = \dfrac{f_{yd}\,(A_s - A'_s)}{b\,d\,f_{cd}}
$$
Sostituendo la relazione di $\psi$ trovata sopra
$$
\frac{16 \xi - 1}{15 \xi}\,\xi = \dfrac{f_{yd}\,(A_s - A'_s)}{b\,d\,f_{cd}}
$$

In [146]:
eq3 = (16*xi-1)/(15*xi)*xi - (fyd*(As-A1s))/(b*d*fcd)
eq3

16*xi/15 - 1/15 - fyd*(-A1s + As)/(b*d*fcd)

In [147]:
sym.solve(eq3, xi)[0]

(-15*A1s*fyd + 15*As*fyd + b*d*fcd)/(16*b*d*fcd)

In [148]:
num_eq3 = eq3.subs({fyd:391.3, As:As_sol1, A1s:A1s_sol1, b:300, d:d_sol1, fcd:14.17})
xi3 = sym.solve(num_eq3, xi)[0]
xi3

0.253453257098057

Risolvendo si ottiene
$$\xi = 0.253453257098057$$

Il limite di $\xi$ per il campo elastico delle armature superiori è 
$$
\xi_{\varepsilon'_{se_{2a/2b}}} = \dfrac{1}{d}\left[\dfrac{\varepsilon'_{se}}{\varepsilon'_{se} + \varepsilon_{su}}\,(d-d'') + d''\right] = \dfrac{1}{460}\left[\dfrac{1.863}{1.863 + 10}\,(460-40) + 40\right] = 0.23034352334074892
$$

In [149]:
sym.simplify((1.863/(1.863+10)) * (d-d2) + d2, rational=True)

78521453258029*d/500000000000000 + 421478546741971*d2/500000000000000

In [150]:
xi3lim_2a2b = ((1.863/(1.863+10)) * (d-d2) + d2).subs({d:d_sol1, d2:40})/d_sol1

In [151]:
print(xi3, '<', xi3lim_2a2b)
xi3 < xi3lim_2a2b

0.253453257098057 < 0.230343523340749


False

Inoltre la deformazione delle armature superiori

In [152]:
print('epsilon1_s =', (10/1000)/(d_sol1 - xi3*d_sol1) * (xi3*d_sol1 - 40), '> epsilon1_se = ',1.863/1000)
(10/1000)/(d_sol1 - xi3*d_sol1) * (xi3*d_sol1 - 40) > 1.863/1000 
#epsilon'_s > epsilon'_se => OK CAMPO 2B

epsilon1_s = 0.00223022519275522 > epsilon1_se =  0.001863


True

Ci si trova effettivamente in campo 2b. Allora
$$
\sigma'_s = f_{yd}
$$

È necessario verificare in che campo di deformazioni è il cls

In [153]:
epsilon_c = (10/1000) * (xi3*d_sol1)/(d_sol1*(1-xi3)) 

print('epsilon_c =', epsilon_c,'< epsilon_c2 = .002')

epsilon_c < 2/1000


#epsilon_c > epsilon_c2 => OK psi calcolato sopra!!!

epsilon_c = 0.00339500854444619 < epsilon_c2 = .002


False

Il cls ha deformazione
$$
\varepsilon_c \in [\varepsilon_{c2}, \varepsilon_{cu}]
$$
perciò i valori di $\psi$ e $\lambda$ calcolati sopra sono corretti.

In [154]:
psi3 = sym.simplify(1-(1/3) * (2/10) * (1-xi)/xi, rational=True).subs({xi:xi3})
psi3

0.803633287534062

In [155]:
mylambda3 = (6*((10/1000) * (xi3*d_sol1)/(d_sol1*(1-xi3)))**2 - 4*((10/1000) * (xi3*d_sol1)/(d_sol1*(1-xi3))) * 2/1000 + (2/1000)**2)/(4*((10/1000) * (xi3*d_sol1)/(d_sol1*(1-xi3)))*(3*((10/1000) * (xi3*d_sol1)/(d_sol1*(1-xi3)))-(2/1000)))
mylambda3

0.413812129258617

Sono verificate le ipotesi di $\varepsilon_c  = 0.344\% \in [\varepsilon_{c2}, \varepsilon_{cu}] = [0.2, 0.35]\%$ e di campo 2b con le armature superiori snervate, essendo $\varepsilon'_{s} = 0.223\% > 0.1863\% = \varepsilon'_{se}$.

Allora, i valori di $\psi$ e $\lambda$ sono
\begin{align}
\psi &= 0.803633287534062\\
\lambda &= 0.413812129258617
\end{align}

In [156]:
MRd = b*psi*xi*d*fcd*(d-mylambda*xi*d) + A1s*fyd*(d-d2)
MRd = MRd.subs({b:300, psi:psi3, xi:xi3, mylambda:mylambda3, d:d_sol1, fcd:14.17, A1s:A1s_sol1, fyd:391.3, d2:40})/10**6
print(MRd, 'kN m')

247.641544543622 kN m


In [157]:
MRd > MEd_N5

True

### La sezione N5 è verificata!

In [158]:
bmin = 6*18+2*10+5*25+2*20
bmin

293

In [159]:
bmin < 300

True

Usando
$$
d = 560\,mm \quad A_s = 6\,\Phi 20\,mm\quad A'_s = 2\,\Phi 20\,mm
$$
la sezione è largamente verificata ma è necessario distribuire le armature metalliche inferiori su due file, oltre ad avere una trave molto alta.

Usando, in alternativa, 
$$
d = 460\,mm \quad A_s = 6\,\Phi 18\,mm \quad A'_s = 2\,\Phi 18\,mm
$$ 
la trave è comunque verificata agli SLU (campo 2b come in precedenza) e inoltre l'armatura può essere contenuta tutta su una fila (in prima approssimazione e salvo calcoli corretti).

<!--È necessario mettere le armature tese su due file!-->

---

## Progetto sezione C1

Si ipotizzi di essere sulla retta limite tra il campo 2 e il campo 3. In questo modo:
$$
\xi_{2/3} = \dfrac{\varepsilon_{cu}}{\varepsilon_{cu} + \varepsilon_{su}}
$$

Assumendo 
\begin{align*}
\varepsilon_{cu} &= 0.35\%\\
\varepsilon_{su} &= 1.0\%
\end{align*}
risulta
$$
\xi_{2/3} = 0.2593
$$

Essendo sulla retta limite tra i campi 2 e 3, i valori di $\psi$ e $\lambda$ sono costanti e valgono
\begin{align*}
\psi &= 0.80952\\
\lambda &= 0.416
\end{align*}

Inoltre sono note le caratteristiche geometriche della sezione
\begin{align}
b &= 300\,mm\\
d &= 460\,mm
\end{align}

Per calcolare le armature si fa l'equilibrio alla rotazione attorno a $T'_s$
$$
-b\,\psi\,\xi\,d\,f_{cd}\,(\lambda\,\xi\,d - d'') + A_s\,f_{yd}\,(d-d'') = M_{Ed}^{C1}
$$

dove 
$$
M_{Ed}^{C1} = 95.1842\,kN\,m
$$

In [160]:
MEd_C1 = 95.1842*10**6

In [161]:
eq = -b*psi*xi*d*fcd*(mylambda* xi *d - d2) + As * fyd *(d-d2) - MEd
eq

As*fyd*(d - d2) - MEd - b*d*fcd*psi*xi*(d*mylambda*xi - d2)

In [162]:
eq = eq.subs({b:300, d:d_sol1, d2:40, fyd:391.3, fcd:14.17, psi:0.80952, mylambda:0.416, xi:.2593, MEd:MEd_C1})
eq_sol = sym.solve(eq, As)[0]
eq_sol

603.195424470401

Si adottano
\begin{align}
A_s &= 3\Phi 18\,mm = 243\pi\,mm^2\\
A'_s &= 2\Phi 18\,mm = 162\pi\,mm^2
\end{align}

In [184]:
As_sol1 = 3*np.pi * 18**2 /4
A1s_sol1 = 2* np.pi *18**2 /4

print(As_sol1, A1s_sol1)

763.4070148223198 508.93800988154646


## Verifica sezione C1
Se ipotizziamo di essere in campo 3, $\psi$ e $\lambda$ sono costanti. Il copriferro superiore al limite elastico tra i campi 2 e 3 vale
$$
d''_{\varepsilon'_{2/3}} = \dfrac{\varepsilon_{cu} - \varepsilon'_{se}}{\varepsilon_{cu}}\,x_{2/3} = \dfrac{0.35-0.1863}{0.35}\%\,\cdot0.2593 \cdot 460\,mm = 55.788\,mm 
$$

Poiché $d'' = 40\,mm < d''_{\varepsilon'_{2/3}} = 55.788\,mm$ le armature superiori snervano in campo 2, sottocampo 2b).

Allora
$$
b\,\psi\,x\,fcd + A'_s\,f'_{yd} - A_s\,f_{yd} = 0
$$

da cui si ricava $x$

In [185]:
eq2 = (16*xi -1)/(15*xi) * xi - (fyd * (As - A1s))/(b*d*fcd)
eq2

16*xi/15 - 1/15 - fyd*(-A1s + As)/(b*d*fcd)

In [186]:
eq2 = eq2.subs({fyd:391.3, A1s:A1s_sol1, As:As_sol1, b:300, d:d_sol1, fcd:14.17})
eq2

16*xi/15 - 0.117587535226149

In [187]:
xi1 = sym.solve(eq2, xi)[0]
xi1

0.110238314274515

$$
\xi = 0.110238314274515 < \xi_{2a/2b} = 0.23034
$$

Ci si trova in campo 2a. **Le armature superiori sono in campo elastico**
$$
\varepsilon'_s = \varepsilon_{su} \,\dfrac{\xi\,d - d''}{d\,(1-\xi)} = 0.026\,\%< 0.1863\,\% = \varepsilon'_{se}
$$
Si deve ricalcolare il valore di $\xi$ corretto da
$$
b\,\psi\,\xi\,d\,f_{cd} + A'_s\,\sigma'_s - A_s\,f_{yd} = 0
$$
cioè
$$
b\,\psi\,\xi\,d\,f_{cd} + A'_s\,E_s\,\varepsilon'_s - A_s\,f_{yd} = 0
$$
dove
$$
\varepsilon'_s = \varepsilon_{su} \,\dfrac{\xi\,d - d''}{d\,(1-\xi)}
$$
In definitiva l'equazione da risolvere è
$$
b\,\dfrac{16\,\xi -1}{15\xi}\,\xi\,d\,f_{cd} + A'_s\,E_s\, \varepsilon_{su} \,\dfrac{\xi\,d - d''}{d\,(1-\xi)} - A_s\,f_{yd} = 0
$$

In [188]:
eq3 = b*((16*xi-1)/(15*xi)) * xi * d * fcd + A1s*Es*(epsilon_su)*((xi*d-d2)/(d*(1-xi)))- As*fyd
eq3

0.01*A1s*Es*(d*xi - d2)/(d*(1 - xi)) - As*fyd + b*d*fcd*(16*xi - 1)/15

In [189]:
eq3 = eq3.subs({b:300, d:d_sol1, As:As_sol1, A1s:A1s_sol1, fyd:391.3, fcd:14.17, Es:210*10**3, epsilon_su:10/1000, d2:40})
eq3

2085824.0*xi - 429085.164899974 + 2323.41265380706*(460*xi - 40)/(1 - xi)

In [190]:
xi2 = sym.solve(eq3, xi)[0]
print(xi2, '< xi_2a2b = 0.23034')
xi2 < 0.23034

0.160696498505634 < xi_2a2b = 0.23034


True

In [191]:
10/1000 * (xi2*d_sol1 - 40)/(d_sol1 - xi2*d_sol1)

0.000878585358397894

In [192]:
epsilon_c2 = 2/1000
epsilon_cu = 3.5/1000
epsilon_su = 10/1000

Ci si trova effettivamente in campo 2a.
$$
\varepsilon'_s = \varepsilon_{su} \,\dfrac{\xi\,d - d''}{d\,(1-\xi)} = 0.08785\% < \varepsilon'_{se}
$$
$$
\dfrac{\varepsilon_{su} + \varepsilon_c}{d} = \dfrac{\varepsilon_{su}}{d(1-\xi)}
$$
da cui
$$
\varepsilon_c = 0.1914\% < \varepsilon_{c2} = 0.2\%
$$
e perciò si deve ricalcolare $\psi$ e $\lambda$ con
$$
\psi = \dfrac{\varepsilon_c}{3\varepsilon_{c2}^2}\,(3\varepsilon_{c2} - \varepsilon_c) = \dfrac{\xi}{1-\xi}\,\dfrac{\varepsilon_{su}}{3\varepsilon_{c2}^2}\,(3\varepsilon_{c2} - \dfrac{\xi}{1-\xi}\,\varepsilon_{su}) 
$$

In [214]:
eq3 = b*xi/(1-xi) * epsilon_su/(3*epsilon_c2**2) * (3*epsilon_c2-(xi/(1-xi))*epsilon_su) * xi*d*fcd + A1s*Es*epsilon_su* ((xi*d-d2)/(d-xi*d)) - As*fyd
eq3 = eq3.subs({b:300, d:d_sol1, fcd:14.17, A1s:A1s_sol1, Es:210*10**3, d2:40, As:As_sol1, fyd:391.3})
xi3 = sym.solve(eq3, xi)[1]
print(xi3 )
xi3 =0.160694044934078
xi3

0.160694044934078 + 0.e-23*I


0.160694044934078

In [223]:
epsilon_c = symbols('epsilon_c')
eq4 = (epsilon_su/(d-xi*d) - (epsilon_su + epsilon_c)/d).subs({d:d_sol1, xi:xi3})
eq4_sol = sym.solve(eq4, epsilon_c)[0]
eq4_sol

0.00191460627634242

In [226]:
mylambda4 = (4*epsilon_c2 - eq4_sol)/(4*(3*epsilon_c2 - eq4_sol))
mylambda4

0.372387224786834

In [231]:
psi4 = (eq4_sol/(3*epsilon_c2**2)) * (3*epsilon_c2 - eq4_sol)
psi4

0.651826705387060

$$
\xi = 0.16069405
$$
$$
\varepsilon_c = 0.19146\%
$$
$$
\lambda = \dfrac{4\varepsilon_{c2} - \varepsilon_c}{4\,(3\varepsilon_{c2} - \varepsilon_c)} = 0.372387224786834
$$
$$
\psi = \dfrac{\varepsilon_c}{3\varepsilon_{c2}^2}\,(3\varepsilon_{c2} - \varepsilon_c) = \dfrac{\xi}{1-\xi}\,\dfrac{\varepsilon_{su}}{3\varepsilon_{c2}^2}\,(3\varepsilon_{c2} - \dfrac{\xi}{1-\xi}\,\varepsilon_{su}) = 0.65182670538706
$$

In [234]:
MRd_C1 = round((-b*psi*xi*d*fcd*(mylambda*xi*d-d2) + As*fyd*(d-d2)).subs({b:300, psi:psi4, xi:xi3, d:d_sol1, fcd:14.17, mylambda:mylambda4, d2:40, As:As_sol1, fyd:391.3}),4)
MRd_C1/10**6

128.0177434799

In [236]:
MRd_C1 > MEd_C1

True

### La sezione C1 è verificata!

---

## Progetto sezione N2

Si ipotizzi di essere sulla retta limite tra il campo 2 e il campo 3. In questo modo:
$$
\xi_{2/3} = \dfrac{\varepsilon_{cu}}{\varepsilon_{cu} + \varepsilon_{su}}
$$

Assumendo 
\begin{align*}
\varepsilon_{cu} &= 0.35\%\\
\varepsilon_{su} &= 1.0\%
\end{align*}
risulta
$$
\xi_{2/3} = 0.2593
$$

Essendo sulla retta limite tra i campi 2 e 3, i valori di $\psi$ e $\lambda$ sono costanti e valgono
\begin{align*}
\psi &= 0.80952\\
\lambda &= 0.416
\end{align*}

Inoltre sono note le caratteristiche geometriche della sezione
\begin{align}
b &= 300\,mm\\
d &= 460\,mm
\end{align}

Per calcolare le armature si fa l'equilibrio alla rotazione attorno a $T'_s$
$$
-b\,\psi\,\xi\,d\,f_{cd}\,(\lambda\,\xi\,d - d'') + A_s\,f_{yd}\,(d-d'') = M_{Ed}^{C1}
$$

dove 
$$
M_{Ed}^{C1} = -196.506\,kN\,m
$$

In [91]:
MEd_N2 = 196.506*10**6

In [92]:
eq = -b*psi*xi*d*fcd*(mylambda* xi *d - d2) + As * fyd *(d-d2) - MEd
eq

As*fyd*(d - d2) - MEd - b*d*fcd*psi*xi*(d*mylambda*xi - d2)

In [93]:
eq = eq.subs({b:300, d:d_sol1, d2:40, fyd:391.3, fcd:14.17, psi:0.80952, mylambda:0.416, xi:.2593, MEd:MEd_N2})
eq_sol = sym.solve(eq, As)[0]
print(eq_sol)
print(eq_sol*.3)

1219.71058151712
365.913174455137


Si adottano
\begin{align}
A_s &= 5\Phi 18\,mm = 405\pi\,mm^2\\
A'_s &= 2\Phi 18\,mm = 162\pi\,mm^2
\end{align}

In [94]:
As_sol1 = round(5*np.pi * 18**2 /4,4)
A1s_sol1 = round(2* np.pi *18**2 /4,4)

print(As_sol1, A1s_sol1)

1272.345 508.938


## Verifica sezione N2
Se ipotizziamo di essere in campo 3, $\psi$ e $\lambda$ sono costanti. Il copriferro superiore al limite elastico tra i campi 2 e 3 vale
$$
d''_{\varepsilon'_{2/3}} = \dfrac{\varepsilon_{cu} - \varepsilon'_{se}}{\varepsilon_{cu}}\,x_{2/3} = \dfrac{0.35-0.1863}{0.35}\%\,\cdot0.2593 \cdot 460\,mm = 55.788\,mm 
$$

Poiché $d'' = 40\,mm < d''_{\varepsilon'_{2/3}} = 55.788\,mm$ le armature superiori snervano in campo 2, sottocampo 2b).

Allora
$$
b\,\psi\,x\,fcd + A'_s\,f'_{yd} - A_s\,f_{yd} = 0
$$

da cui si ricava $x$

In [95]:
eq2 = (16*xi -1)/(15*xi) * xi - (fyd * (As - A1s))/(b*d*fcd)
eq2

16*xi/15 - 1/15 - fyd*(-A1s + As)/(b*d*fcd)

In [96]:
eq2 = eq2.subs({fyd:391.3, A1s:A1s_sol1, As:As_sol1, b:300, d:d_sol1, fcd:14.17})
eq2

16*xi/15 - 0.219429269379072

In [97]:
xi1 = sym.solve(eq2, xi)[0]
xi1

0.205714940042880

$$
\xi = 0.20571494004288 < \xi_{2a/2b} = 0.23034
$$

Ci si trova in campo 2a. **Le armature superiori sono in campo elastico**
$$
\varepsilon'_s = \varepsilon_{su} \,\dfrac{\xi\,d - d''}{d\,(1-\xi)} = 0.1495\,\%< 0.1863\,\% = \varepsilon'_{se}
$$
Si deve ricalcolare il valore di $\xi$ corretto da
$$
b\,\psi\,\xi\,d\,f_{cd} + A'_s\,\sigma'_s - A_s\,f_{yd} = 0
$$
cioè
$$
b\,\psi\,\xi\,d\,f_{cd} + A'_s\,E_s\,\varepsilon'_s - A_s\,f_{yd} = 0
$$
dove
$$
\varepsilon'_s = \varepsilon_{su} \,\dfrac{\xi\,d - d''}{d\,(1-\xi)}
$$
In definitiva l'equazione da risolvere è
$$
b\,\dfrac{16\,\xi -1}{15\xi}\,\xi\,d\,f_{cd} + A'_s\,E_s\, \varepsilon_{su} \,\dfrac{\xi\,d - d''}{d\,(1-\xi)} - A_s\,f_{yd} = 0
$$

In [102]:
eq3 = b*((16*xi-1)/(15*xi)) * xi * d * fcd + A1s*Es*(epsilon_su)*((xi*d-d2)/(d*(1-xi)))- As*fyd
eq3

A1s*Es*epsilon_su*(d*xi - d2)/(d*(1 - xi)) - As*fyd + b*d*fcd*(16*xi - 1)/15

In [103]:
eq3 = eq3.subs({b:300, d:d_sol1, As:As_sol1, A1s:A1s_sol1, fyd:391.3, fcd:14.17, Es:210*10**3, epsilon_su:10/1000, d2:40})
eq3

2085824.0*xi - 628232.5985 + 2323.41260869565*(460*xi - 40)/(1 - xi)

In [104]:
xi2 = sym.solve(eq3, xi)[0]
print(xi2, '< xi_2a2b = 0.23034')
xi2 < 0.23034

0.216484182907451 < xi_2a2b = 0.23034


True

In [106]:
10/1000 * (xi2*d_sol1 - 40)/(d_sol1 - xi2*d_sol1)

#epsilon1_s

0.00165315949394574

In [111]:
epsilon_c = (10/1000)*(xi2*d_sol1)/(d_sol1*(1-xi2))

#epsilon_c

In [116]:
psi_2 = (16*xi2-1)/(15*xi2)
psi_2

0.758714991993199

In [115]:
mylambda_2 = ((6*epsilon_c**2 - 4*epsilon_c*2/1000 + (2/1000)**2)/(4*epsilon_c*(3*epsilon_c-2/1000)))
mylambda_2

0.398540738618410

Ci si trova effettivamente in campo 2a.
$$
\varepsilon'_s = \varepsilon_{su} \,\dfrac{\xi\,d - d''}{d\,(1-\xi)} = 0.165316\% < \varepsilon'_{se}
$$
$$
\dfrac{\varepsilon_{su} + \varepsilon_c}{d} = \dfrac{\varepsilon_{su}}{d(1-\xi)}
$$
da cui
$$
\varepsilon_c = 0.2763\% > \varepsilon_{c2} = 0.2\%
$$
e perciò si deve ricalcolare $\psi$ e $\lambda$ con
$$
\psi = 1 - \dfrac{1}{3}\,\dfrac{\varepsilon_{c2}}{\varepsilon_{c}} = 1 - \dfrac{1}{3}\,\dfrac{\varepsilon_{c2}}{\varepsilon_{su}}\,\dfrac{1-\xi}{\xi} = 1 - \dfrac{1}{3}\,\dfrac{2}{10}\cdot\dfrac{1-\xi}{\xi} = \frac{16 \xi - 1}{15 \xi} = 0.758714991993199
$$
$$
\lambda = \dfrac{6\varepsilon_c^2 - 4\varepsilon_c\,\varepsilon_{c2} + \varepsilon_{c2}^2}{4\varepsilon_c\,(3\varepsilon_c - \varepsilon_{c2}} =0.39854073861841
$$

In [128]:
MRd_N2 = -b*psi*xi*d*fcd*(mylambda*xi*d-d2) + As*fyd*(d-d2)
MRd_N2 = round(MRd_N2.subs({b:300, psi:psi_2, xi:xi2, d:d_sol1, fcd:14.17, mylambda:mylambda_2, d2:40, As:As_sol1, fyd:391.3}),4)
MRd_N2/10**6

209.2050938395

In [129]:
MRd_N2 > MEd_N2

True

### La sezione N2 è verificata!
---